hibernate-core/src/main/java/org/hibernate/id/IdentifierGeneratorHelper.java

```diff
@@ -134,32 +138,45 @@ else if ( clazz == BigDecimal.class ) {
 			}
 		}
 		else {
-			if ( clazz == Long.class ) {
-				return rs.getLong( identifier );
-			}
-			else if ( clazz == Integer.class ) {
-				return rs.getInt( identifier );
-			}
-			else if ( clazz == Short.class ) {
-				return rs.getShort( identifier );
-			}
-			else if ( clazz == String.class ) {
-				return rs.getString( identifier );
+			try {
+				return extractIdentifier( rs, identifier, type, clazz );
 			}
-			else if ( clazz == BigInteger.class ) {
-				return rs.getBigDecimal( identifier ).setScale( 0, BigDecimal.ROUND_UNNECESSARY ).toBigInteger();
-			}
-			else if ( clazz == BigDecimal.class ) {
-				return rs.getBigDecimal( identifier ).setScale( 0, BigDecimal.ROUND_UNNECESSARY );
-			}
-			else {
-				throw new IdentifierGenerationException(
-						"unrecognized id type : " + type.getName() + " -> " + clazz.getName()
-				);
+			catch (SQLException e) {
+				if(Identifier.isQuoted( identifier )) {
```

Comment:
Unless I am mistaken... that identifier, if quoted, would be surrounded by the Dialect's quoting chars.  `Identifier#isQuoted` is checking the "source form" or quoting (with one additional pattern check).  In other words, `Identifier#isQuoted` will match \`name\`, [name] and "name".  But it would not match, e.g., 'name'.  I think it might be better to check here using `Dialect#openQuote()` and `Dialect#closeQuote()`